libs/langchain-community/src/vectorstores/firestore.ts

```diff
@@ -0,0 +1,234 @@
+
+    private readonly metadataKey: string;
+
+    private readonly vectorType: SupportedVectorTypes;
+
+    private caller: AsyncCaller;
+
+    _vectorstoreType(): string {
+        return 'firestore';
+    }
+
+    constructor(
+        embeddings: EmbeddingsInterface,
+        args: FirestoreVectorSearchLibArgs
+    ) {
+        super(embeddings, args);
+        this.collection = args.collection;
+        this.textKey = args.textKey ?? 'text';
+        this.embeddingKey = args.embeddingKey ?? 'embedding';
+        this.metadataKey = args.metadataKey ?? 'metadata';
+        this.vectorType = args.vectorType ?? 'COSINE'
+        this.caller = new AsyncCaller(args);
+    }
+
+    /**
+     * Method to add vectors and their corresponding documents to Firestore
+     * collection.
+     * @param vectors Vectors to be added.
+     * @param documents Corresponding documents to be added.
+     * @returns Promise that resolves when the vectors and documents have been added.
+     */
+    async addVectors(
+        vectors: number[][],
+        documents: Document[],
+        options?: { ids?: string[] }
+    ) {
+        const docs = vectors.map((embedding, idx) => ({
+            [this.textKey]: documents[idx].pageContent,
+            [this.embeddingKey]: FieldValue.vector(embedding),
+            [this.metadataKey]: documents[idx].metadata,
+        }));
+        if (options?.ids === undefined) {
+            const batch = this.collection.firestore.batch()
+            const newDocIds: string[] = [];
+
+            // Add new documents to the batch
+            docs.forEach(doc => {
+                const newDocRef = this.collection.doc(); // Generate a new document reference
+                batch.set(newDocRef, doc);
+                newDocIds.push(newDocRef.id); // Keep track of the generated document IDs
+            });
+
+            await batch.commit();
+            return newDocIds;
+        } else {
+            if (options.ids.length !== vectors.length) {
+                throw new Error(
+                    'If provided, "options.ids" must be an array with the same length as "vectors".'
+                );
+            }
+            const batch = this.collection.firestore.batch()
+            const { ids } = options;
+            for (let i = 0; i < docs.length; i += 1) {
+                const docRef = this.collection.doc(ids[i]);
+                // Update the fields you want
+                batch.update(docRef, docs[i]);
+            }
+            await batch.commit()
+            return options?.ids
+        }
+    }
+
+    /**
+     * Method to add documents to the Firestore collection. It first converts
+     * the documents to vectors using the embeddings and then calls the
+     * addVectors method.
+     * @param documents Documents to be added.
+     * @returns Promise that resolves when the documents have been added.
+     */
+    async addDocuments(documents: Document[], options?: { ids?: string[] }) {
```

Comment:
For document input arguments, the type should be `DocumentInterface`, and for return types, `Document`.
```suggestion
    async addDocuments(documents: DocumentInterface[], options?: { ids?: string[] }) {
```